dir.c

```diff
@@ -2731,7 +2731,7 @@ static struct untracked_cache_dir *validate_untracked_cache(struct dir_struct *d
 	}
 
 	if (!dir->untracked->root) {
-		const int len = sizeof(*dir->untracked->root);
+		const int len = st_add(sizeof(*dir->untracked->root), 1);
 		dir->untracked->root = xmalloc(len);
 		memset(dir->untracked->root, 0, len);
```

Comment:
I think the intent here is that the `name` flex-array would be the empty string (your "account for the flex array" confused me for a moment; I think the issue is that the array is supposed to be a NUL-terminated string, so we are really accounting for that).

So I think your "extend the `memset`" in the commit message is also "extend the `malloc`". We need a place to put that NUL byte (and what the patch is doing here is correct).

I'm likewise curious that this would show up on Windows and not elsewhere. It would be a non-problem on any system that defines `FLEX_ARRAY` to 1. But most compilers (especially modern Unix ones like gcc or clang) will not. So I'd expect them to be vulnerable above all.

I do think this whole thing could be written as:

```
FLEX_ALLOC_STR(dir->untracked->root, name, "");
```

That IMHO shows the intent more clearly, as well as getting rid of the manual computation (which is the reason I added those macros in the first place; I didn't catch this in my audits because it erroneously wasn't doing any computation in the first place!). It also avoids the use of `int` for storing a computed length (it's OK here because we know the name field is the empty string, but it would be likely to come up in a grep-based audit).

Reply:
Thanks!  I was hoping there was a better way to write that.  (I must admit that after several hours of wandering aimlessly thru the debugger and finally finding the problem, I didn't think to look for a FLEX_ macro to do it better.)